### PR TITLE
Implement EncryptionKeyRotation controllers

### PR DESCRIPTION
**Pull request by black-dragon74**
This patch implements the sidecar service for
EncryptionKeyRotation and registers it

A new option is added to csi-addons test tool
to test the EncryptionKeyRotation operation

### CODE REVIEW

**Comment by black-dragon74** [comment]
Note to reviewer: This PR will cover everything related to implementing EncryptionKeyRotation.

I have separated changes into separate commits to make it easy to review. I will squash all the commits before merging.

Regards

File: cmd/csi-addons/encryptionkeyrotation.go
Comment (Madhu-1) [inline cmd/csi-addons/encryptionkeyrotation.go:59]: what is the expected response from `ekrRes`?

File: internal/proto/encryptionkeyrotation.proto
Comment (Madhu-1) [inline internal/proto/encryptionkeyrotation.proto:7]: `operation on the pv`

File: internal/proto/encryptionkeyrotation.proto
Comment (Madhu-1) [inline internal/proto/encryptionkeyrotation.proto:13]: can you please reword this one as it wont be sent to SP rather its sent to the addons sidecar and it process the request

File: internal/proto/encryptionkeyrotation.proto
Comment (Madhu-1) [inline internal/proto/encryptionkeyrotation.proto:23]: remove empty line

File: internal/proto/encryptionkeyrotation.proto
Comment (Madhu-1) [inline internal/proto/encryptionkeyrotation.proto:24]: GitHub is complaining about `end line`

File: internal/sidecar/service/encryptionkeyrotation.go
Comment (Madhu-1) [inline internal/sidecar/service/encryptionkeyrotation.go:31]: please group the imports as required like https://github.com/ceph/ceph-csi/blob/devel/docs/coding.md#imports

File: internal/sidecar/service/encryptionkeyrotation.go
Comment (Madhu-1) [inline internal/sidecar/service/encryptionkeyrotation.go:63]: return `err` as part of error as well

File: internal/sidecar/service/encryptionkeyrotation.go
Comment (Madhu-1) [inline internal/sidecar/service/encryptionkeyrotation.go:79]: This check seems wrong, if check is for `NodeStageSecretRef` but we are doing get operation for NodePublish secrets

File: internal/sidecar/service/encryptionkeyrotation.go
Comment (Madhu-1) [inline internal/sidecar/service/encryptionkeyrotation.go:81]: logs not log here as we are doing it anywhere in above error states

File: internal/sidecar/service/encryptionkeyrotation.go
Comment (Madhu-1) [inline internal/sidecar/service/encryptionkeyrotation.go:93]: i think this seems redundant we dont need to check `ekrResult` as we are not using it, error check seems to be more than enough here.

File: internal/proto/encryptionkeyrotation.proto
Comment (Madhu-1) [inline internal/proto/encryptionkeyrotation.proto:9]: as its a proto file lets not use short form. use `Encryption` instead of `Enc` in all other places in proto.

File: cmd/csi-addons/encryptionkeyrotation.go
Comment (Madhu-1) [inline cmd/csi-addons/encryptionkeyrotation.go:60]: add a log that operation is successful with required details here at the end

**Review by Madhu-1 — CHANGES_REQUESTED** [review]

File: cmd/csi-addons/encryptionkeyrotation.go
Comment (black-dragon74) [reply cmd/csi-addons/encryptionkeyrotation.go:59]: And empty instance of `*proto.EncKeyRotateResponse`

File: internal/sidecar/service/encryptionkeyrotation.go
Comment (black-dragon74) [reply internal/sidecar/service/encryptionkeyrotation.go:81]: Can you please reiterate?

File: internal/proto/encryptionkeyrotation.proto
Comment (black-dragon74) [reply internal/proto/encryptionkeyrotation.proto:9]: I did that to avoid the proto namespace collision. It cannot be Encryption else it will conflict with the ns in spec.

Any other alternatives?

Comment (Madhu-1) [reply internal/proto/encryptionkeyrotation.proto:9]: can you please explain a bit on conflict with the NS, i might be missing something.

File: internal/sidecar/service/encryptionkeyrotation.go
Comment (Madhu-1) [reply internal/sidecar/service/encryptionkeyrotation.go:80]: i meant to not log here just remove it.

File: internal/sidecar/service/encryptionkeyrotation.go
Comment (Madhu-1) [inline internal/sidecar/service/encryptionkeyrotation.go:90]: `key for volume` to `key for pv`

File: internal/proto/encryptionkeyrotation.proto
Comment (black-dragon74) [reply internal/proto/encryptionkeyrotation.proto:9]: If I name the service `EncryptionKeyRotation` it will be the same as the service name in the imported spec from `cis-addons/spec`. It will compile fine but will panic at the run time due to confilicting namespaces, i.e. both of them (`internal/proto` and `csi-addons/spec`) having the same member names `EncryptionKeyRotation`.

Such collision is not allowed in protobuf.

Comment (Madhu-1) [reply internal/proto/encryptionkeyrotation.proto:9]: EncryptionKeyRotationController is the serviceName in the csi-addons/spec and here it will be EncryptionKeyRotation, isn't it?

Comment (black-dragon74) [reply internal/proto/encryptionkeyrotation.proto:9]: The above statement was just an example. This applies to message names as well.

`EncryptionKeyRotateRequest` and `EncryptionKeyRotateResponse` are the conflicting members.

**Comment by black-dragon74** [comment]
Note: This PR has been rebased to https://github.com/black-dragon74/kubernetes-csi-addons/tree/migrate-go3 as kubebuilder (latest version) doesn't allow creation of resources with old layout.

Once https://github.com/csi-addons/kubernetes-csi-addons/pull/606 is merged, I will rebase it with main.

Regards

**Comment by nixpanic** [comment]
@black-dragon74 quite a few conflicts need to be addressed now :disappointed:

**Comment by black-dragon74** [comment]
> @black-dragon74 quite a few conflicts need to be addressed now 😞

I was expecting it 😂 . Taking a look 

Regards

**Comment by black-dragon74** [comment]
I have tested the `csi-addons` tool along with the controllers for key rotation. Everything works as expected. The support for RWX volumes shall be added later (see commit msg).

Regards

File: cmd/csi-addons/encryptionkeyrotation.go
Comment (Madhu-1) [inline cmd/csi-addons/encryptionkeyrotation.go:38]: `persistent volume` to `PersistentVolume` use errors.New()

File: cmd/csi-addons/encryptionkeyrotation.go
Comment (Madhu-1) [reply cmd/csi-addons/encryptionkeyrotation.go:60]: This is still not addressed yet

File: cmd/manager/main.go
Comment (Madhu-1) [inline cmd/manager/main.go:220]: use `defaultTimeout`

File: internal/controller/csiaddons/encryptionkeyrotationcronjob_controller.go
Comment (Madhu-1) [inline internal/controller/csiaddons/encryptionkeyrotationcronjob_controller.go:2]: 2024

File: internal/controller/csiaddons/encryptionkeyrotationcronjob_controller.go
Comment (Madhu-1) [inline internal/controller/csiaddons/encryptionkeyrotationcronjob_controller.go:36]: adjust the import order

File: internal/controller/csiaddons/encryptionkeyrotationcronjob_controller_test.go
Comment (Madhu-1) [inline internal/controller/csiaddons/encryptionkeyrotationcronjob_controller_test.go:2]: 2024

File: internal/controller/csiaddons/encryptionkeyrotationcronjob_controller_test.go
Comment (Madhu-1) [inline internal/controller/csiaddons/encryptionkeyrotationcronjob_controller_test.go:29]: adjust the order

File: internal/controller/csiaddons/encryptionkeyrotationjob_controller.go
Comment (Madhu-1) [inline internal/controller/csiaddons/encryptionkeyrotationjob_controller.go:2]: 2024

File: internal/controller/csiaddons/encryptionkeyrotationjob_controller.go
Comment (Madhu-1) [inline internal/controller/csiaddons/encryptionkeyrotationjob_controller.go:44]: adjust order

File: internal/controller/csiaddons/encryptionkeyrotationjob_controller.go
Comment (Madhu-1) [inline internal/controller/csiaddons/encryptionkeyrotationjob_controller.go:57]: remove extra space

File: internal/controller/csiaddons/encryptionkeyrotationjob_controller_test.go
Comment (Madhu-1) [inline internal/controller/csiaddons/encryptionkeyrotationjob_controller_test.go:2]: 2024

File: internal/controller/csiaddons/encryptionkeyrotationjob_controller_test.go
Comment (Madhu-1) [inline internal/controller/csiaddons/encryptionkeyrotationjob_controller_test.go:30]: adjust the order

File: internal/sidecar/service/encryptionkeyrotation.go
Comment (Madhu-1) [inline internal/sidecar/service/encryptionkeyrotation.go:73]: we cannot add this check, the error should come from csi driver if it wont support it

File: internal/sidecar/service/encryptionkeyrotation.go
Comment (Madhu-1) [inline internal/sidecar/service/encryptionkeyrotation.go:85]: can you check this one, if node stage secret is missing we need to use publish secret

**Review by Madhu-1 — CHANGES_REQUESTED** [review]
major follow up refractoring is required as we have more duplication between this and the reclaimspace

File: cmd/csi-addons/encryptionkeyrotation.go
Comment (black-dragon74) [reply cmd/csi-addons/encryptionkeyrotation.go:38]: Can we use capitalized error statements?

Comment (Madhu-1) [reply cmd/csi-addons/encryptionkeyrotation.go:38]: lets try and see if CI complains, if not we can have it as it is

Comment (black-dragon74) [reply cmd/csi-addons/encryptionkeyrotation.go:38]: It is a golang convention to have error strings start with lowercase and without period. I do not think we have checks in CI for this though.

File: internal/sidecar/service/encryptionkeyrotation.go
Comment (black-dragon74) [reply internal/sidecar/service/encryptionkeyrotation.go:85]: Both are optional values, since `NodeStageVolume` comes before `NodePublishVolume` in the order of things, it would be present in the cases where it is required. 

Can you tell me of such case where stage secret might be empty but publish secret wouldn't?

File: internal/controller/csiaddons/encryptionkeyrotationjob_controller.go
Comment (black-dragon74) [reply internal/controller/csiaddons/encryptionkeyrotationjob_controller.go:57]: Removed. I would like to add that the newest version of kubebuilder now adds a space by default.

File: api/csiaddons/v1alpha1/encryptionkeyrotationcronjob_types.go
Comment (Madhu-1) [inline api/csiaddons/v1alpha1/encryptionkeyrotationcronjob_types.go:2]: 2024

File: api/csiaddons/v1alpha1/encryptionkeyrotationjob_types.go
Comment (Madhu-1) [inline api/csiaddons/v1alpha1/encryptionkeyrotationjob_types.go:2]: 2024

File: config/samples/csiaddons_v1alpha1_encryptionkeyrotationcronjob.yaml
Comment (Madhu-1) [inline config/samples/csiaddons_v1alpha1_encryptionkeyrotationcronjob.yaml:13]: lets not use rbd, name this as example-pvc

File: config/samples/csiaddons_v1alpha1_encryptionkeyrotationjob.yaml
Comment (Madhu-1) [inline config/samples/csiaddons_v1alpha1_encryptionkeyrotationjob.yaml:10]: lets not use rbd, name this as example-pvc

File: internal/controller/csiaddons/encryptionkeyrotationcronjob_controller.go
Comment (Madhu-1) [inline internal/controller/csiaddons/encryptionkeyrotationcronjob_controller.go:36]: change the order of import, internal should be 2nd in list and last should be the third party imports

File: internal/controller/csiaddons/encryptionkeyrotationjob_controller.go
Comment (Madhu-1) [inline internal/controller/csiaddons/encryptionkeyrotationjob_controller.go:44]: change the order of import

File: internal/sidecar/service/encryptionkeyrotation.go
Comment (Madhu-1) [reply internal/sidecar/service/encryptionkeyrotation.go:73]: @black-dragon74 this need to be addressed.

**Review by Madhu-1 — COMMENTED** [review]
@black-dragon74 please fix the CI issue

File: cmd/csi-addons/encryptionkeyrotation.go
Comment (yati1998) [reply cmd/csi-addons/encryptionkeyrotation.go]: we can add error strings with upper case, it just shows some indications in code base, but good to maintain the convention as used elsewhere in the project.
So, in my opinion it's fine to use `PersistentVolume`

File: cmd/csi-addons/encryptionkeyrotation.go
Comment (yati1998) [inline cmd/csi-addons/encryptionkeyrotation.go:59]: `enc` -> `encryption`

**Comment by black-dragon74** [comment]
> @black-dragon74 please fix the CI issue

It (git diff) is failing as the protoc has been updated to 3.20.2 and the code in the current main branch used 3.19.x

IG this is expected.

**Comment by Madhu-1** [comment]
> > @black-dragon74 please fix the CI issue
> 
> It (git diff) is failing as the protoc has been updated to 3.20.2 and the code in the current main branch uses 3.19.x
> 
> IG this is expected.

We can't merge a PR if CI fails, please update the CI version here https://github.com/csi-addons/kubernetes-csi-addons/blob/1a1dd5f6e469f99eeb2f55ad18069f83698657c7/.github/workflows/test-golang.yaml#L27 to match the expectation

File: internal/proto/encryptionkeyrotation_grpc.pb.go
Comment (nixpanic) [inline internal/proto/encryptionkeyrotation_grpc.pb.go:4]: this is an unrelated change, there should be a commit that updates the protoc version and updated generated files.

File: internal/proto/networkfence.pb.go
Comment (nixpanic) [inline internal/proto/networkfence.pb.go:4]: this is an unrelated change, there should be a commit that updates the protoc version and updated generated files.

**Review by nixpanic — COMMENTED** [review]
It would be much easier to review if you can split this huge commit into several ones:
- changes to `api/`
- changes to `PROJECT` and other related `config/` files
- implementation of the controller
- implementation of the sidecar
- implementation of the csi-addons testing too
- documentation + example under `doc/` directory (**missing completely**)
- ... maybe others?

**Comment by black-dragon74** [comment]
> It would be much easier to review if you can split this huge commit into several ones:

It was exactly like that (split into 7 commits). I will split it again :(



> * documentation + example under `doc/` directory (**missing completely**)

Will add it in a future PR.

**Review by Rakshith-R — COMMENTED** [review]
LGTM

File: config/rbac/kustomization.yaml
Comment (black-dragon74) [reply config/rbac/kustomization.yaml:33]: Yes. Done

**Review by nixpanic — APPROVED** [review]
Looks much better to me, thanks!

I have not checked the reconcile in very much detail, I'll leave it up to @Rakshith-R to check it.

**Review by Madhu-1 — APPROVED** [review]

**Comment by nixpanic** [comment]
@mergifyio rebase

**Pull request merged**
